lib/Conversion/TritonToTritonGPU/TritonToTritonGPU.cpp

```diff
@@ -223,7 +223,7 @@ struct TritonDotPattern : public OpConversionPattern<triton::DotOp> {
                   ConversionPatternRewriter &rewriter) const override {
     RankedTensorType origType = op.getType().cast<RankedTensorType>();
     auto origShape = origType.getShape();
-    auto typeConverter = cast<TritonGPUTypeConverter>(getTypeConverter());
+    auto typeConverter = static_cast<TritonGPUTypeConverter*>(getTypeConverter());
```

Comment:
Is this bypass the `classof` check? Seems not safe.

Reply:
Same comment here. Isn't `cast` better than `static_cast` here?

Reply:
@ptillet 
Yes, I agree. I've changed to `auto typeConverter = getTypeConverter<TritonGPUTypeConverter>()`, which leverages the MLIR interface and looks better.

However, the official implement of this interface is:
```cpp
   template <typename ConverterTy>
   std::enable_if_t<std::is_base_of<TypeConverter, ConverterTy>::value,
                    ConverterTy *>
   getTypeConverter() const {
     return static_cast<ConverterTy *>(typeConverter);
   }
```
, which is unsafe too.

The reason I guess is that MLIR does not implement the typeinfo (e.g. `TypeID` for the `TypeConverter` class). It seems a non-polymorphic class, where the only way to cast is static.

Another solution would be to rewrite a `classof` function like this and still use `cast`:
```cpp
class TritonGPUTypeConverter : public TypeConverter {
public:
  static bool classof(const TypeConverter *typeConverter) {
    return true;
  }
}
```

But "any `TypeConverter` is a `TritonGPUTypeConverter`" seems unsafe too. I guess changing to the MLIR interface is enough, and we may wait for MLIR to update its implementation :)